delete_job functionality

gpetretto opened the pull request:
A new option to delete a single Job of a Flow.
No checks for consistency are performed, so it is the user's responsibility to leave the Flow in a consistent state.

gpetretto committed:
delete_job functionality

davidwaroquiers commented on src/jobflow_remote/utils/db.py:
```diff
@@ -142,11 +142,11 @@ def __init__(
         self.lock_id = lock_id or suuid()
         self.kwargs = kwargs
         self.update_on_release: dict | list = {}
+        self.delete_on_release: bool = False
```
Maybe add to the docstring in the main characteristics that it allows to delete the document upon release, and add an example (similar to update_on_release). Maybe one thing that could be good (even if it is more verbose, but I had to look into the code to know how to to an update on release) is to add an "Attributes" section in the class docstring (somehow in "addition" to the __init__ "Parameters" section), so that update_on_release and delete_on_release (maybe a few others that are not in the arguments ?) are clearly documented. What do you think ?

davidwaroquiers replied:
Would also be good to add some tests just for MongoLock's update_on_release and delete_on_release (outside of its usage elsewhere). I can do that in another PR.

davidwaroquiers commented on src/jobflow_remote/utils/db.py:
```diff
@@ -243,21 +243,36 @@ def release(self, exc_type, exc_val, exc_tb) -> None:
+
+        if self.delete_on_release:
+            if self.update_on_release:
+                raise ValueError(
```
Maybe this check could be done when trying to set update_on_release and/or delete_on_release ? It would require to modify the code a bit. Probably make update_on_release and delete_on_release properties with getter/setter methods and having this check in the setter methods. This way the raise is not at release time (which then may leave the document in a locked state because it never gets to the update of locks to None). What do you think ?

gpetretto replied:
Seems a good idea.

davidwaroquiers commented on src/jobflow_remote/cli/job.py:
```diff
@@ -494,6 +494,61 @@ def stop(
+    verbosity: verbosity_opt = 0,
+    wait: wait_lock_opt = None,
+    raise_on_error: raise_on_error_opt = False,
+    delete_output: Annotated[
```
This is the same as the one in flows. Should it be shared ? (and go in types.py)

davidwaroquiers commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -3189,6 +3190,7 @@ def update_flow_state(
         updated_states
             A dictionary with the updated states of Jobs that have not been
             stored in the DB yet. In the form {job_uuid: JobState value}.
+            If the value is None the Job is considered deleted.
```
Just to make sure I understand properly, the value of the state of a deleted Job might be None while doing the delete (and inside the lock_job_flow  context in the delete_job method) right ? But a Job can never have a None state in the database correct ? Maybe it might be good to explain this as a comment in the delete_job method (and here). I was initially "puzzled" by the fact that a job state could be None and I was thinking then are these (or this, in this case) jobs still in the database but "set" as "deleted" ("using" a None state).

gpetretto replied:
The Job is never really in the None state. The point here is just to signal that the state of the Job should be ignored, since it will be deleted in the update. I will try to clarify the usage.

davidwaroquiers commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -3570,6 +3576,190 @@ def remove_batch_process(self, process_id: str, worker: str) -> dict:
+
+            # Remove job from jobs list if no job with that id remains in the flow
+            if not any(job_uuid == id_tuple[1] for id_tuple in flow_doc.ids):
+                flow_doc.jobs.remove(job_uuid)
```
Maybe I'm not enough familiar with the structure of the FlowDoc but I thought the jobs list had multiple occurrences of the same job uuid if there has been some replace. If that's the case shouldn't it always remove the job_uuid ? (i.e. it would remove one of them). Then related to that, if it's the case, is there a matching between the jobs list and the ids list ? Maybe I just missed the fact that the jobs attribute in FlowDoc contains the unique uuids.

gpetretto replied:
My idea was that the `jobs` attribute should only contain one entry for each uuid, but I see that indeed when a replace is present it contains the same uuid multiple times. Aside from this specific case it should not cause issues, since it should always be used just to find a Flow that contains a Job with a specific uuid. So it does not matter that it is present multuple times. Nonetheless I would consider this a bug and fix it. 
In that case this would be correct, but maybe I will change this to 
```python 
flow_doc.jobs = [jid for jid in flow_doc.jobs if jid != job_uuid]
```
to be backward compatible.
What do you think?

davidwaroquiers commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -3570,6 +3576,190 @@ def remove_batch_process(self, process_id: str, worker: str) -> dict:
+            for parent_dict in flow_doc.parents.values():
+                for index_list in parent_dict.values():
+                    if job_uuid in index_list:
+                        index_list.remove(job_uuid)
```
What if there has been a replace and we delete the job with index 2 ? Shouldn't the job with index 1 still be a parent ? Maybe I'm also missing something in the logic.

gpetretto replied:
Yes. Thanks for spotting this!

davidwaroquiers commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -3570,6 +3576,190 @@ def remove_batch_process(self, process_id: str, worker: str) -> dict:
+                        stacklevel=2,
+                    )
+
+            return [job_doc["db_id"]]
```
Why does it return a list ? (it's actually also a question for all other kinds of job-related methods)

gpetretto replied:
Because all the methods acting on multiple jobs use the `_many_jobs_action` to apply the same action on a list of jobs and call the methods that act on a single job. This one expects a list of jobs as output. I thought it would be easier to keep the same API for all these methods (also in the beginning there were much less such mehtods and the first one was `rerun_job`, that indeed return a list of multiple ids). However it would also be possible to handle this in `_many_jobs_action` and let methods that only modify one job to return a single entry. Do you think it would be better?
In the case of `delete_job` it may be dubious, since maybe in the future we may introduce the option to also delete linked jobs?

davidwaroquiers reviewed (commented):
Thanks for this!
I have some comments, questions and maybe thoughts for improvements but maybe some (or most/all :p) of these are because I am not fully aware of everything in the code, and I did not have time to completely go through the rest and fully grasp all the details.

gpetretto committed:
update MongoLock and add tests. Address review comments

gpetretto committed:
optionally delete files  when deleting jobs

gpetretto committed:
improve options to prevent accidental deletion of flows and jobs

gpetretto committed:
fix test

gpetretto commented:
Some changes with the last commits.
* added the option to delete the files when deleting the Job.
* moved the logic to delete the files in `JobController` (also when deleting Flows)
* changed the signature of the functions that deal with a single job in `JobController`: now if the function acts only on a single Job returns the single db_id, not a list with a single element. The change is handled in `_many_jobs_action`.
* Fix for #169 by allowing to set the maximum number of flows to be deleted.
* add the same option for job deletion as well.

davidwaroquiers reviewed (approved):
Great work! Thanks!
